### ISSUE TITLE
Remove `Array.set_at`

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#4524]
This task is automatically imported from the old Task Issue Board and it was originally created by James Dunkerley.
Original issue is [here](https://www.pivotaltracker.com/story/show/182879865).
<hr />

In order to make Enso more immutable.

Want to change `Vector.new` to use a `Vector.builder` with a prefixed size and then the `append` function.

For Vector.builder: 
- We should an ArrayList instead of the growing array.
- Then we just use the `append` method.
- `to_array` should not exist on the builder.
- `to_vector` becomes a method which gets the Vector.
    - So we would copy the ArrayList to an Array via Array.copy.
    - As a copy we don't need to builder
- `append_vector source start=0 length=source.length` append a sub range of another vector
    - We want to use Array.copy under the hood.
- If we can't do with ArrayList then roll our own...

- Remove `Array.set_at`

We rely on `Vector.slice` for `append_vector` - so we will ensure that it works without additional overhead when we implement efficient slices in https://www.pivotaltracker.com/story/show/182880080 . For now, it will do a bit more copying.

## Comments:
**Radosław Waśko** reports a new **STANDUP** for today (2022-08-04):

**Progress:** Applying code review suggestions to Vector.take+drop. Meetings. Initial implementation for removing Array.set_at. It should be finished by 2022-08-10.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/182879865). (After I'm back from timeoff) Test the set_at changes, test a hypothesis if we can make Enso Array a bit more immutable, how it will play with our usages of System.arraycopy (Enso Bot - Aug 4, 2022)
<hr />
**Radosław Waśko** reports a new **🔴 DELAY** for today (2022-08-10):

**Summary:** There is **1 days** delay in implementation of the [Remove `Array.set_at` (#182879865) task](https://www.pivotaltracker.com/story/show/182879865).
It **will cause 0 days delay** for the delivery of this weekly plan.

**Delay Cause:** I've got caught up in figuring out a weird imports bug. It has gone away after I cleared the caches but I wanted to see if I can repro it and so I did. It took some time though. Also I encountered a very weird and unexpected behavior when modifying the Array.set_at - now some tests that used to work (and I don't really know how it worked in the past...) stopped working - which is not too surprising, but now I will need to figure out how to deal with this in a sensible manner. In short - the way JSON serialization works seems to slightly abuse how we use atoms. For some reason it used to work before though, so I will see if it is possible to fix it somehow. (Enso Bot - Aug 10, 2022)
<hr />
**Radosław Waśko** reports a new **STANDUP** for today (2022-08-10):

**Progress:** Mostly catching up. Reading PRs. Got stuck on an imports bug for a while, managed to find a repro and reported it on pivotal. Also reported a task we discussed regarding stacktraces. Got caught up on my Array.set_at work. It should be finished by 2022-08-11.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/182879865). Figure out how to fix the broken tests, test a hypothesis if we can make Enso Array a bit more immutable, how it will play with our usages of System.arraycopy (Enso Bot - Aug 10, 2022)
<hr />
**Radosław Waśko** reports a new **🔴 DELAY** for yesterday (2022-08-24):

**Summary:** There is **14 days** delay in implementation of the [Remove `Array.set_at` (#182879865) task](https://www.pivotaltracker.com/story/show/182879865).
It **will cause 0 days delay** for the delivery of this weekly plan.

**Delay Cause:** The task was put on hold because it was blocked by another task. (Enso Bot - Aug 25, 2022)
<hr />
**Radosław Waśko** reports a new **STANDUP** for yesterday (2022-08-24):

**Progress:** Prepared the Table.take+drop and Array.set_at PRs for integration. Starting on next tasks (will focus on Datetime storage). Examined a small task on date formats that is worth fixing quickly. It should be finished by 2022-08-25.

**Next Day:** Next day I will be working on the [##183033133 task](https://www.pivotaltracker.com/story/show/183033133). Implement a quickfix for Data_Formatter and add a few more tests, continue work on Storage. (Enso Bot - Aug 25, 2022)
<hr />
